Bluetooth: host: directed advertising support

kapi-no opened the pull request:
Closes #8437

Added implementation to the directed advertising API in the Connection
Management module. Introduced a new connection state for this type of
advertising. The new state is symmetric to the connection state used for
scanning.

Added new advertising options that can be used to trigger low and high
duty directed advertising. Extended the Advertising Parameters
descriptor, so now it includes peer address. Added macros for default
values of Advertising Parameters, which are used to trigger directed
advertising.

Signed-off-by: Kamil Piszczek <Kamil.Piszczek@nordicsemi.no>

jhedberg commented on include/bluetooth/bluetooth.h:
```diff
@@ -271,6 +271,12 @@ enum {
+	BT_LE_ADV_OPT_DIR_MODE_LOW_DUTY = BIT(4),
+
+	/* Use high duty directed advertising mode */
+	BT_LE_ADV_OPT_DIR_MODE = BIT(5),
```
We only have 8 bits to spare for the options, so I think it'd be better to use  `peer_addr != NULL` to know we're doing directed and then a single flag to toggle between low or high duty.

jhedberg commented on include/bluetooth/bluetooth.h:
```diff
@@ -286,36 +292,55 @@ struct bt_le_adv_param {
+  * @param _peer_addr Peer address, used only for directed advertising mode
   */
-#define BT_LE_ADV_PARAM(_options, _int_min, _int_max) \
+#define BT_LE_ADV_PARAM(_options, _int_min, _int_max, _peer_addr) \
```
This breaks all existing applications using this macro, so if we can avoid changing it that'd be best.

Vudentz replied:
We should probably have a different macro for direct adv.

jhedberg commented on include/bluetooth/conn.h:
```diff
@@ -227,13 +227,11 @@ int bt_le_set_auto_conn(bt_addr_le_t *addr,
  */
-struct bt_conn *bt_conn_create_slave_le(const bt_addr_le_t *peer,
-					const struct bt_le_adv_param *param);
+struct bt_conn *bt_conn_create_slave_le(const struct bt_le_adv_param *param);
```
Why are you moving the peer_addr into the options? Is it because you're calling `bt_le_adv_start()` internally and that API does not take a peer_addr parameter? In that case I'd refactor bt_le_adv_start() to a public and an internal function. The public stays as-is and the internal one takes a peer_addr (which can be NULL). This also eliminates the issue with breaking the BT_LE_ADV_OPT() macro for existing users.

kapi-no replied:
That's one reason. Apart from that, peer address is considered to be an Advertising Parameter according to the Bluetooth specification (like ```interval_min```, ```interval_max```), so I thought it would be okay to put it in the ```bt_le_adv_param``` structure.

I don't have any strong opinion on this, so I will go along with your suggestion

jhedberg replied:
I understand your reasoning. However, it'd be nice not to impose the 4 bytes overhead on most users of the API who don't use directed advertising. Also, having the address in the adv_param would raise the question how `bt_le_adv_start()` should behave with a non-NULL peer_addr if called directly by the app instead of through bt_conn_create_slave_le(). I think it's something we'd want to prevent, i.e. it should probably have returned an error. Even with the new design I think `bt_le_adv_start()` should return an error if called with the new advertising option set.

jhedberg replied:
If you refactor the code to an internal `static int adv_start(...)` function that both `bt_conn_create_slave_le()` and `bt_le_adv_start()` call, then then `bt_le_adv_start()` could check for the flag before calling the internal function. Telling the app it did something wrong instead of silently ignoring an explicitly given option seems like more helpful policy toward the developer. Another alternative could be an assert.

Btw, with the single-bit approach, I'd make the default high-duty and call the new option LOW_DUTY, simply because high-duty came first in the core spec versions.

jhedberg commented on subsys/bluetooth/controller/ll_sw/ctrl.c:
```diff
@@ -6208,7 +6208,7 @@ static void mayfly_adv_stop(void *param)
 	radio_le_conn_cmplt = (void *)pdu_data_rx->lldata;
 	memset(radio_le_conn_cmplt, 0x00, sizeof(struct radio_le_conn_cmplt));
-	radio_le_conn_cmplt->status = 0x3c;
+	radio_le_conn_cmplt->status = BT_HCI_ERR_ADV_TIMEOUT;
```
This is a good fix, but it should be in a separate commit.

cvinayak replied:
Rebase, I think this change is in upstream as part of my other PR.

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -823,7 +838,12 @@ static void le_enh_conn_complete(struct bt_hci_evt_le_enh_conn_complete *evt)
 	 * CONNECT state associated with passed peer LE address.
 	 */
-	conn = bt_conn_lookup_state_le(&id_addr, BT_CONN_CONNECT);
+	for (u32_t i = 0; i < ARRAY_SIZE(le_states); i++) {
```
Same here. Feel free to refactor into a small helper function if you think that's better.

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -5060,53 +5090,63 @@ int bt_le_adv_start(const struct bt_le_adv_param *param,
+						== BT_DATA_NAME_COMPLETE ||
+					    sd[i].type
+						== BT_DATA_NAME_SHORTENED) {
+						return -EINVAL;
```
Ugh.. this deep indentation causes some really ugly line splits. Could you try to redesign this somehow, there are many possibilities like inverted conditions, labels, helper functions, etc).

jhedberg reviewed (changes requested):
Thanks for taking the time to implement this missing feature! There are a few issues I'd like to get sorted out before merging however (see my inline comments).

cvinayak commented on subsys/bluetooth/host/conn.c:
```diff
@@ -1770,11 +1779,18 @@ int bt_conn_disconnect(struct bt_conn *conn, u8_t reason)
+			/* User should unref connection object when receiving
+			 * error in connection callback.
+			 */
+			return bt_le_adv_stop();
```
Are you stopping advertising on disconnect of master role?

jhedberg replied:
This looks suspicious to me as well. Shouldn't the bt_le_adv_stop() only be done for BT_CONN_CONNECT_DIR_ADV? Now it's done also for BT_CONN_CONNECT_SCAN, which can't be right, ban it? Remember that the app may have both CONFIG_BT_PERIPHERAL=y and CONFIG_BT_CENTRAL=y.

cvinayak reviewed (changes requested):
Once Johan's comments are addressed I will continue review

kapi-no committed:
Bluetooth: host: directed advertising support

Added implementation to the directed advertising API in the Connection
Management module. Introduced a new connection state for this type of
advertising. The new state is symmetric to the connection state used for
scanning.

Added a new advertising option that can be used to trigger low and high
duty directed advertising. Added macros for default values of
Advertising Parameters, which are used to trigger directed advertising.

Signed-off-by: Kamil Piszczek <Kamil.Piszczek@nordicsemi.no>

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -5034,79 +5057,100 @@ static bool valid_adv_param(const struct bt_le_adv_param *param)
-int bt_le_adv_start(const struct bt_le_adv_param *param,
-		    const struct bt_data *ad, size_t ad_len,
-		    const struct bt_data *sd, size_t sd_len)
+static inline bool name_is_present_in_adv_data(const struct bt_data *ad,
```
I'd just call this `ad_has_name()`

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -5034,79 +5057,100 @@ static bool valid_adv_param(const struct bt_le_adv_param *param)
+		    ad[i].type
+			== BT_DATA_NAME_SHORTENED) {
+			return true;
+		}
```
```
                for (i = 0; i < ad_len; i++) {
                        if (ad[i].type == BT_DATA_NAME_COMPLETE ||
                            ad[i].type == BT_DATA_NAME_SHORTENED) {
                                return true;
                        }
                }
```

kapi-no replied:
Yes, sorry. I forgot about the main reason why I created this helper function

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -765,6 +765,24 @@ static void update_pending_id(struct bt_keys *keys, void *data)
+	 * Make lookup to check if there's a connection object in
+	 * CONNECT or DIR_ADV state associated with passed peer LE address.
+	 */
+	for (u32_t i = 0; i < ARRAY_SIZE(le_states); i++) {
```
Maybe I was unclear, but I meant that using a for-loop (even if you have just one) seems overkill if you always have the same fixed two iterations. Doing two if-statements is (at least for me) more readable to understand what the code is doing since you don't then have to look to the beginning of the function to figure out what the contents of the le_states array is. Or do you think it's likely that this array will be extended in the future? Another issue: I don't think the Zephyr coding style allows mixing code and declarations, i.e. variable declarations (such as `i`) should be in the beginning of a code block.

pdunaj replied:
>>> I don't think the Zephyr coding style allows mixing code and declarations, i.e. variable declarations (such as i) should be in the beginning of a code block.

@jhedberg I don't think there is any rule for that. No document requires that or forbids this c99 feature. People just don't do it. If I am wrong please point to the document - I could not find any for Zephyr or for Linux kernel.
I am asking as mixing code and variables greatly improves readability and reduces maintenance costs.
Of course we could limit ourselves and do it but I don't see absolutely any reason for it. It was requited by old compilers that would not be able to cope with such mixing but we are well into 21st century now...

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -765,6 +765,24 @@ static void update_pending_id(struct bt_keys *keys, void *data)
 }
 #endif
 
+static struct bt_conn *find_valid_conn(bt_addr_le_t *peer_addr)
```
This is not looking up any valid (ref-count > 0) connection, so this name is a bit misleading. Would `find_pending_connect()` be more appropriate?

jhedberg commented on subsys/bluetooth/host/hci_core.h:
```diff
@@ -198,3 +198,7 @@ void bt_id_del(struct bt_keys *keys);
+
+int adv_start(const struct bt_le_adv_param *param, const struct bt_data *ad,
+	      size_t ad_len, const struct bt_data *sd, size_t sd_len,
+	      const bt_addr_le_t *peer);
```
Why is this here? Just make it a static function in hci_core.c (move it high enough so you don't need to do forward declarations for it)

kapi-no replied:
I need to access this function in bt_conn_create_slave_le() (in the conn.c file), so it cannot be static

jhedberg replied:
Right, sorry, I forgot that conn.c has the directed advertising function. Then we should add a bit of name-spacing to the function's name. E.g. everything in hci_core.h right now is at least prefixed with bt_ so we'll want to be consistent with that. The obvious name is already taken by the public API. `bt_le_adv_start_internal()` perhaps? (not really liking the length of it)

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -765,6 +765,24 @@ static void update_pending_id(struct bt_keys *keys, void *data)
+			break;
+		}
+	}
+	return conn;
```
Minor coding style thing: empty line after { (unless the next line is also })

jhedberg commented on subsys/bluetooth/host/conn.c:
```diff
@@ -1912,7 +1928,36 @@ int bt_le_set_auto_conn(bt_addr_le_t *addr,
+		bt_conn_unref(conn);
+		return NULL;
+	}
+	bt_conn_set_state(conn, BT_CONN_CONNECT_DIR_ADV);
```
Minor coding style thing: Empty line after {

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -789,13 +807,22 @@ static void le_enh_conn_complete(struct bt_hci_evt_le_enh_conn_complete *evt)
-		conn = bt_conn_lookup_state_le(NULL, BT_CONN_CONNECT);
+		conn = find_valid_conn(NULL);
+
 		if (!conn) {
```
Minor coding style thing: No empty line between assignment to a variable and testing the variable's value.

kapi-no commented:
I have updated this PR according to @jhedberg suggestions

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -765,6 +765,24 @@ static void update_pending_id(struct bt_keys *keys, void *data)
+
+	conn = bt_conn_lookup_state_le(peer_addr, BT_CONN_CONNECT_DIR_ADV);
+
+	return conn;
```
Minor: this could just be `return bt_conn_lookup_state_le(peer_addr, BT_CONN_CONNECT_DIR_ADV);`

jhedberg commented on include/bluetooth/bluetooth.h:
```diff
@@ -310,6 +316,16 @@ struct bt_le_adv_param {
+			BT_GAP_ADV_FAST_INT_MIN_2, BT_GAP_ADV_FAST_INT_MAX_2)
+
+#define BT_LE_ADV_CONN_DIR BT_LE_ADV_PARAM(BT_LE_ADV_OPT_CONNECTABLE | \
+					   BT_LE_ADV_OPT_ONE_TIME, 0, 0)
```
Should BT_LE_ADV_OPT_ONE_TIME be implied by doing directed advertising? What if someone forgets to set it? I have a feeling that the places that are checking `if (ONE_TIME)` should be changed to `if (ONE_TIME || directed)`.

kapi-no replied:
Yes, I can add guard in ```valid_adv_params``` function, so it is not possible to start directed advertising without ```ONE_TIME``` option set

jhedberg replied:
I think I'd prefer the implied approach, i.e. bt_le_adv_start_internal() would set BT_DEV_KEEP_ADVERTISING if either ONE_TIME was set or a non-NULL peer_addr was given.

cvinayak replied:
Have the options BT_LE_ADV_OPT_CONNECTABLE and BT_LE_ADV_OPT_ONE_TIME implicit, API bt_conn_create_slave_le does not need these to be set.

jhedberg reviewed (changes requested):
Looks much better now, thanks a lot! There are however still a few things that need fixing (unless I misunderstood something with the code). Also, could you try adding a third patch to this PR which adds directed advertising support to the Bluetooth shell? That way we can easily test and verify that the code works.

jhedberg commented:
> Could you point me to any sample that uses bt_shell?

I use mostly `tests/bluetooth/shell` however the actual code you'll need to look at (and the file you'll need to modify) is `subsys/bluetooth/shell/bt.c`.

jhedberg commented:
This actually raises an interesting question: how have you been testing the code so far? :)

kapi-no commented:
> This actually raises an interesting question: how have you been testing the code so far? :)

I modified one Bluetooth sample application and used it to test this feature. I wasn't aware that there is a better way of doing it (with shell)

kapi-no committed:
Bluetooth: shell: adding command for directed advertising

Added a new command, which can be used to test the directed advertising
API. This command allows user to turn on directed advertising in two
modes: low and high duty. High duty mode is the default one.

Signed-off-by: Kamil Piszczek <Kamil.Piszczek@nordicsemi.no>

kapi-no commented:
Added a command to bt shell, that allows user to start/stop directed advertising in the high/low duty mode

jhedberg commented on subsys/bluetooth/shell/bt.c:
```diff
@@ -994,6 +1003,64 @@ static int cmd_auto_conn(int argc, char *argv[])
+	}
+
+	if (default_conn) {
+		printk("Pointer to the connection instance is not free\n");
```
Would it not make sense to return an error here, otherwise you create a memory leak of a bt_conn object?

kapi-no replied:
Yes, that's true - I forgot about it. However, I suggest to just return 0 at this point, because then the user won't get this command usage information (the command is correct in this case, but its not possible to start advertising because of application state).

jhedberg reviewed (changes requested):
The shell patch looks pretty good. Just one small comment.

kapi-no commented:
@jhedberg Applied fixes from your latest review. Any new comments/suggestions?

kapi-no commented:
Yes, it should simplify this command a bit.

>I'm not sure the PR can go in as it's a feature and the merge window has closed

That's okay. When is the next merge window then?

kapi-no commented:
Now, the ```directed-adv``` command is very similar to the ```connect``` command.

cvinayak commented on subsys/bluetooth/host/conn.c:
```diff
@@ -1776,6 +1787,16 @@ int bt_conn_disconnect(struct bt_conn *conn, u8_t reason)
 			bt_le_scan_update(false);
 		}
 		return 0;
+	case BT_CONN_CONNECT_DIR_ADV:
```
Directed Advertising depends on Peripheral Role, all code here needs to be compiled out, not just the bt_le_adv_stop().

kapi-no replied:
It is handled in a similar way as the ```BT_CONN_CONNECT_SCAN``` state. Should I align both cases?

cvinayak commented on subsys/bluetooth/host/conn.c:
```diff
@@ -1912,7 +1933,37 @@ int bt_le_set_auto_conn(bt_addr_le_t *addr,
+		return NULL;
+	}
+
+	bt_conn_set_state(conn, BT_CONN_CONNECT_DIR_ADV);
```
Should we set the state before starting advertising? (in case a connection is established quickly before we reach here?

kapi-no replied:
Well, I didn't encounter any problems with it while testing  (even in the test scenario, where high duty advertising is triggered when there is a connect request already pending from central), but I think it is worth changing just for safety even though it would require state clean up in case of an error from ```bt_le_adv_start_internal```

cvinayak replied:
ignore this. No action required.

cvinayak commented on subsys/bluetooth/shell/bt.c:
```diff
@@ -2173,6 +2211,7 @@ static const struct shell_cmd bt_commands[] = {
 	{ "connect", cmd_connect_le, HELP_ADDR_LE },
 	{ "disconnect", cmd_disconnect, HELP_NONE },
 	{ "auto-conn", cmd_auto_conn, HELP_ADDR_LE },
+	{ "directed-adv", cmd_directed_adv, HELP_ADDR_LE " <mode: low>" },
```
use "[" and "]"  for optional parameter?

cvinayak commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -5189,7 +5238,7 @@ int bt_le_adv_start(const struct bt_le_adv_param *param,
 	}
 
-	if (!(param->options & BT_LE_ADV_OPT_ONE_TIME)) {
+	if (!(param->options & BT_LE_ADV_OPT_ONE_TIME) && !dir_adv) {
```
if BT_LE_ADV_OPT_ONE_TIME is set implicitly when dir_adv is assigned to true, we dont need the additional !dir_adv check.

jhedberg replied:
Problem with this is that we allow `param` to be const, i.e. we'd need to have a local copy of it (probably not worth  it)

kapi-no replied:
If we really want to take this implicit approach, it is also possible to make a copy of ```param``` in the ```bt_conn_create_slave_le``` function and set proper options

kapi-no replied:
So what is the consensus over this ?
1) leave it as it is
2) copy ```param``` to the local variable in the ```bt_le_adv_start_internal``` and set the implicit options if it is directed advertising mode
3) copy ```param``` to the local variable in the ```bt_conn_create_slave_le ``` and set the implicit options

jhedberg replied:
I'd go for 1, since it's quite easy to add the extra condition to `bt_le_adv_start_internal()` and `valid_adv_param()`, same way as was done for ONE_TIME.

cvinayak replied:
`param->options` is referenced 8 times (indirect addressing) in this function, I will go for an auto variable `options = param->options;`, `if (dir_adv) { options |= (...);} /* set one time and connectable flags */`, and use `options` in the other 8 places.

jhedberg replied:
I'm ok for bt_le_adv_start_internal(). As for valid_adv_param() I think it'd be overkill there - you don't have the same 8 references etc that @cvinayak was talking about.

jhedberg replied:
That said, another option could be to create the auto variable higher up the stack, i.e. immediately in the call from the application, and then pass it together with the param to start_internal and valid_adv_param.

jhedberg replied:
Mainly just brainstorming here without more carefully looking at the resulting code - do what feels good and I'll take a look at the next iteration of your patch :)

cvinayak replied:
> That said, another option could be to create the auto variable higher up the stack, i.e. immediately in the call from the application, and then pass it together with the param to start_internal and valid_adv_param.

@kapi-no give a try on this suggestion, having upto 4 max. parameters does not degrade performance, and the parameter serves as an auto-variable anyway.

kapi-no replied:
> 3. copy param to the local variable in the bt_conn_create_slave_le and set the implicit options

Isn't it the same as my initial suggestion?

jhedberg replied:
Not quite - I think the idea is to copy the options, not the full parameters struct

jhedberg replied:
..not that there's a big difference

cvinayak commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -5034,79 +5054,99 @@ static bool valid_adv_param(const struct bt_le_adv_param *param)
 
-	d[0].data = ad;
-	d[0].len = ad_len;
+	if (!dir_adv) {
```
in the else clause you can set the connectable and one time options.

kapi-no commented:
Updated PR

kapi-no commented on subsys/bluetooth/shell/bt.c:
```diff
@@ -994,6 +994,44 @@ static int cmd_auto_conn(int argc, char *argv[])
 	return 0;
 }
 
+static int cmd_directed_adv(int argc, char *argv[])
```
Hmm, I could do that, but I think that the BR addresses will not be handled correctly in this case. 

Alternatively, I could add module variable to detect if connected callback is called for the directed advertising connection object and then act accordingly.

Or I can populate ```conn->le.init_addr``` field in ```bt_conn_create_slave_le()```

kapi-no commented on subsys/bluetooth/host/conn.c:
```diff
@@ -1912,7 +1933,43 @@ int bt_le_set_auto_conn(bt_addr_le_t *addr,
+	param_int.options |= (BT_LE_ADV_OPT_CONNECTABLE |
+			      BT_LE_ADV_OPT_ONE_TIME);
+
+	err = bt_le_adv_start_internal(&param_int, NULL, 0, NULL, 0, peer);
```
I didn't consider this corner case - I will look into it

kapi-no commented:
Applied fixes, suggested by @joerchan

jhedberg commented:
@sjanc @Vudentz any comments on this PR?

kapi-no commented:
Can we finalize this PR ? @cvinayak any comments after your last review?

cvinayak commented on subsys/bluetooth/host/conn.c:
```diff
@@ -1912,7 +1933,54 @@ int bt_le_set_auto_conn(bt_addr_le_t *addr,
+	struct bt_conn *conn;
+	struct bt_le_adv_param param_int;
+
+	memcpy(&param_int, param, sizeof(*param));
```
sizeof(param_int), use destination's size. I know the size is same, but good to follow a convention.

cvinayak commented on subsys/bluetooth/host/conn.c:
```diff
@@ -1912,7 +1933,54 @@ int bt_le_set_auto_conn(bt_addr_le_t *addr,
+			err = bt_le_adv_start_internal(&param_int, NULL, 0,
+						       NULL, 0, peer);
+			if (err && (err != -EALREADY)) {
+				BT_WARN("Directed advertising could no be"
```
"no" -> "not"

cvinayak commented on subsys/bluetooth/host/conn.c:
```diff
@@ -1912,7 +1933,54 @@ int bt_le_set_auto_conn(bt_addr_le_t *addr,
+
+	err = bt_le_adv_start_internal(&param_int, NULL, 0, NULL, 0, peer);
+	if (err) {
+		BT_WARN("Directed advertising could no be started: %d", err);
```
"no" -> "not"

jhedberg commented:
For the record, from my side this looks like it's ready to be merged. Just waiting for @cvinayak to be ok with it.

cvinayak reviewed (approved):
LGTM
